mixpanel_oauth_to_redshift.go

```diff
@@ -0,0 +1,177 @@
+
+	for k := range params {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	joined_args := ""
+	for _, k := range keys {
+		joined_args += k
+		value, err := StringOrJsonMarshal(params[k])
+		if err != nil {
+			return "", err
+		}
+		joined_args += fmt.Sprintf("=%s", value)
+	}
+
+	hash := md5.Sum([]byte(joined_args + m.api_secret))
+	return hex.EncodeToString(hash[:]), nil
+}
+
+func (m *MixpanelExport) Request(method string, params map[string]interface{}) ([]byte, error) {
+	params["api_key"] = m.api_key
+	expire_time := time.Now().Add(*mixpanel_export_request_timeout)
+	params["expire"] = expire_time.Unix()
+	params["format"] = "json"
+	delete(params, "sig")
+	sig, err := m.ComputeSig(params)
+	if err != nil {
+		return nil, err
+	}
+	params["sig"] = sig
+
+	values := url.Values{}
+	for key, value := range params {
+		str_value, err := StringOrJsonMarshal(value)
+		if err != nil {
+			return nil, err
+		}
+		values.Set(key, str_value)
+	}
+	url := fmt.Sprintf("%s/%s/%s/?%s", m.endpoint, m.version, method, values.Encode())
+	log.Println(url)
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	return ioutil.ReadAll(resp.Body)
+}
+
+func CopyJsonDataFromS3ToRedshift(
+	db *sql.DB, redshift_table string, export_file string, aws_access_key_id string,
+	aws_secret_access_key string, jsonpaths_file string, aws_region string) error {
+	copy_cmd := fmt.Sprint(
+		"COPY ", redshift_table, " FROM '", export_file, "' WITH ",
+		"credentials ", fmt.Sprintf("'%s=%s;%s=%s'", "aws_access_key_id", aws_access_key_id, "aws_secret_access_key", aws_secret_access_key),
+		" json '", jsonpaths_file,
+		"' region '", aws_region,
+		"' timeformat 'epochsecs'")
+	log.Print("Executing command: ", copy_cmd)
+	_, err := db.Exec(copy_cmd)
+	return err
+}
+
+func main() {
+	flag.Parse()
+	export_file := fmt.Sprintf("%s/%s", *mixpanel_export_dir, *mixpanel_export_date)
+
+	if *export_from_mixpanel {
+		mixpanel_export := NewMixpanelExport()
+		log.Println("Exporting mixpanel data for", *mixpanel_export_date)
+		params := map[string]interface{}{
+			"event":     [1]string{"authorize"},
+			"from_date": *mixpanel_export_date,
+			"to_date":   *mixpanel_export_date,
+		}
+		body, err := mixpanel_export.Request("export", params)
+		if err != nil {
+			log.Panic(err)
+		}
+		err = pathio.Write(export_file, body)
+		if err != nil {
+			log.Panic(err)
+		}
+	}
+
+	if *copy_to_redshift {
+		redshift_source := fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=%s",
+			redshift_host, redshift_port, redshift_user, redshift_password, redshift_database)
+		log.Println("Connecting to Reshift Source: ", redshift_source)
+		db, err := sql.Open("postgres", redshift_source)
+		if err != nil {
+			log.Panic(err)
+		}
+		defer db.Close()
+		err = CopyJsonDataFromS3ToRedshift(
+			db, *redshift_table, export_file, aws_access_key_id, aws_secret_access_key, *jsonpaths_file, aws_region)
+		if err != nil {
+			log.Panic(err)
+		}
+	}
+}
```

Comment:
We need some way of determining whether the database has the data for a particular day or not. One way to do that is  to store a state in some file whenever a copy command for a particular date finishes. These improvements, along with others can come in a follow up PR, possibly along with the change to get analytics DB into redshift.